man/man1/orbintv.1
.TH ORBINTV 1NEMO "10 February 2004"

.SH "NAME"
orbintv \- integrating single stellar orbit with variable timestep

.SH "SYNOPSIS"
.PP
\fBorbintv in=\fPorbit \fBout=\fPorbit [parameter=value]

.SH "DESCRIPTION"
\fBorbintv\fP will integrate a stellar orbit using a higher order
variable timestep. A number of constant timestepping algorithms are implemented
in \fIorbint(1NEMO)\fP.
.PP
Currently two choices for the integratore can be made:
\fBmode=dopri5\fP uses an explicit Runge-Kutta method of
order (4)5 due to Dormand & Prince with step size control and dense output.
\fBmode=dop853\fP uses an explicit Runge-Kutta method of
order 8(5,3) due to Dormand & Prince with step size control and dense output.
.PP
Non-rotating potentials are also supported, but only with the
rotation axis defined along the Z-axis.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fIin-file\fP
input file, in \fIorbit(5NEMO)\fP format [no default]
.TP
\fBout=\fIout-file\fP
output file, will be in \fIorbit(5NEMO)\fP format [no default]
.TP
\fBdtout=\fP
Output timestep.
.TP
\fBtstop=\fP
Stop time [default: \fB10\fP].
.TP
\fBdt=\fItime-step\fP
initial guess for the time step to be taken.
[default: \fB0.01\fP].
.TP
\fBndiag=\fInumber\fP
number of output timesteps when diagnostics are checked and output
to \fIstderr\fP. Use 0 when never.
[default: \fB0\fP].
.TP
\fBpotname=\fIname\fP
name of file of \fIpotential(5NEMO)\fP descriptor. If no name
given, the potential name of the input orbit is used.
[default: none].
.TP
\fBpotpars=\fIpar-list\fP
List of parameters to the potential descriptor. The first
parameter MUST be the pattern speed in the x-y plane.
The remaining parameters are used by the
\fI_inipotential()\fP routine in the potential descriptor.
[default: none - use default from input orbit].
.TP
\fBpotfile=\fIname\fP
name of an optional datafile to the potential descriptor.
This might be an N-body snapshot or list of spline fit
coefficients etc. [default: none - use default from input
orbit].
.TP
\fBmode=\fIint_mode\fP
Specify the integration mode. Any one of \fBdopri5\fP,
\fBdop853\fP
[Default: \fBdopri5\fP].
.TP
\fBtol=\fP
Tolerance used by the integrators to control time stepping.
Negative numbers will cause this number to raised to the
power 10.
[Default: -7]

.SH "EXAMPLES"
The following example launches a particle from the Y axis (at y=1)
in the X direction (speed 0.4) in a plummer potential. Although
the 6D initial conditions are fully specified, so a potential
(potname=) is not needed, it is tagged along with the orbit, such
that the orbint integrator will use it. The integrated orbit is
then passed on to a simple plotting program, which plots an X-Y 
view of this 2D orbit.
.EX

mkorbit - x=0 y=1 z=0 vx=0.4 vy=0 vz=0 potname=plummer |\\
     orbintv - - tstop=10 dt=0.05 |\\
     orbplot -
Using pattern speed = 0
pos: 0.000000 1.000000 0.000000  
vel: 0.400000 0.000000 0.000000  
etot: -0.627107
lz=-0.400000
Pattern speed=0

.EE

.SH "CAVEAT"
For large values of the output timestep, energy conservation may not seem
very good. This is a result of the dense output interpolator.

.SH "SEE ALSO"
mkorbit(1NEMO), orblist(1NEMO), orbint(1NEMO), potential(5NEMO), newton0(1NEMO)
.PP

.SH "AUTHOR"
.nf
Peter Teuben (NEMO adaptation)
Hairer & G. Wanner (Fortran version of core integrators)
J.Colinge (C versions of the core integrators)
.fi

.SH "FILES"
.nf
.ta +2.5i
src/orbit/misc  	orbintv.c, dopri5.c, dop853.c
.fi
.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
15-may-11	V1.0: created	PJT
.fi


src/orbit/misc/orbintv.c
/*
 *  ORBINTV: integrate stellar orbits with variable timestep
 *           for the harder problems
 *
 *      15-may-2011    Cloned off orbint               Peter Teuben
 *      11-jun-2016    fixed *** stack smashing detected ***
 *
 * @todo   implement ORBIT_PHI
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <orbit.h>

#include <dopri5.h>
#include <dop853.h>

string defv[] = {
  "in=???\n		input filename (an orbit) ",
  "out=???\n		output filename (an orbit) ",
  "dtout=0.1\n          Output step time",
  "tstop=10\n           Stop time",
  "dt=0.01\n            (initial) timestep (not used)",
  "ndiag=0\n		** frequency of diagnostics output (0=none)",
  "potname=\n	  	potential name (default from orbit)",
  "potpars=\n	        parameters of potential ",
  "potfile=\n		extra data-file for potential ",
  "mode=dopri5\n        integration method (dopri5 dop853)",
  "tol=-7\n             tolerance of integration",
  "VERSION=1.1b\n       18-apr-2026 PJT",
  NULL,
};

string usage = "integrate stellar orbits with variable timestep";


#ifndef HUGE
#  define HUGE 1.0e20
#endif

string	infile, outfile;		/* file names */
stream  instr, outstr;			/* file streams */

orbitptr o_in  = NULL;			/* pointer to input orbit */
orbitptr o_out = NULL;			/* pointer to output orbit */

int    nsteps, ndiag;          		/* how often */
real   dt, dtout;			/* stepping */
real   tstop;                           /* stop time */
real   omega, omega2, tomega;  		/* pattern speed */
real   tdum=0.0;                        /* time used in potential() */
real   eta ;                            /* tolerance */
bool   Qomega;                          /* if omega != 0 */



extern int match(string, string, int *);


proc pot;				/* pointer to the potential */

void integrate_dopri5(void),
     integrate_dop853(void),
     setparams(void), 
     prepare(void);
real print_diag(double time, double *posvel);


/*----------------------------------------------------------------------------*/
void nemo_main ()
{
    int imode;

    setparams();
							       /* open files */
    instr = stropen (infile,"r");

    get_history(instr);			
    if (read_orbit(instr,&o_in)==0)   	 /* read input orbit */
		error ("error in reading input orbit");
    strclose(instr);
    /* replace default orbit with user supplied, if given */
    if (hasvalue("potname")) PotName(o_in) = getparam("potname");
    if (hasvalue("potpars")) PotPars(o_in) = getparam("potpars");
    if (hasvalue("potfile")) PotFile(o_in) = getparam("potfile");

    if (allocate_orbit (&o_out,Ndim(o_in),nsteps)==0)
		error ("Error allocating output orbit");
    pot=get_potential(PotName(o_in), PotPars(o_in), PotFile(o_in));
    if (pot==NULL) 
		error("Potential %s could not be loaded",PotName(o_in));
    PotName(o_out) =  PotName(o_in);
    PotPars(o_out) =  PotPars(o_in);
    PotFile(o_out) =  PotFile(o_in);

    omega = get_pattern();
    dprintf(0,"Pattern speed=%g\n",omega);
    omega2 = omega*omega;
    tomega = 2.0*omega;
    Qomega = omega != 0.0;

    outstr = stropen (outfile,"w");
    prepare();
    match(getparam("mode"),"dopri5 dop853 end",&imode);
    if (imode==0x01)
        integrate_dopri5();
    else if (imode==0x02)
        integrate_dop853();
    else
        error("imode=0x%x; Illegal integration mode=",imode);

    put_history(outstr);
    write_orbit (outstr,o_out); 		/* write output file */
    strclose(outstr);
}

void setparams(void) 
{
    infile = getparam("in");
    outfile = getparam("out");

    nsteps = 1000;
    dt = getdparam("dt");
    dtout = getdparam("dtout");
    tstop = getdparam("tstop");
    nsteps = (int)(tstop/dtout + 1.001);
    dprintf(0,"nsteps = %d\n",nsteps);

    ndiag=getiparam("ndiag");
    eta = getdparam("tol");
    if (eta < 0) 
      eta = pow(10.0,eta);
}

void prepare(void)
{
    Masso(o_out) = Masso(o_in);
}

/* helper functions for the integrator */

static double last_pot;

void rhs(unsigned n, double x, double *y, double *f)
{

  double pos[3], vel[3], acc[3], time;
  int ndim = 3;

  if (n!=6) error("not 3D");

  pos[0] = y[0];
  pos[1] = y[1];
  pos[2] = y[2];
  vel[0] = y[3];
  vel[1] = y[4];
  vel[2] = y[5];
  time = x;
  (*pot)(&ndim,pos,acc,&last_pot,&time);
  if (Qomega) {
    acc[0] += omega2*pos[0] + tomega*vel[1];    /* rotating frame */
    acc[1] += omega2*pos[1] - tomega*vel[0];    /* corrections    */
  }

  f[0] = y[3];
  f[1] = y[4];
  f[2] = y[5];
  f[3] = acc[0];
  f[4] = acc[1];
  f[5] = acc[2];

}

void solout5(long nr, double xold, double x, double *y, unsigned n, int *irtrn)
{
  static double xout;
  static int isave;
  double pv[6];

  if (nr==1) {
    xout = x + dtout;
    isave = 0;
    Torb(o_out,isave) = x;
    Xorb(o_out,isave) = y[0]; Yorb(o_out,isave) = y[1]; Zorb(o_out,isave) = y[2];
    Uorb(o_out,isave) = y[3]; Vorb(o_out,isave) = y[4]; Worb(o_out,isave) = y[5];
    (void) print_diag(x, y);
  } else  {
    while (x >= xout) {
      isave += 1;
      Torb(o_out,isave) = xout;
      pv[0] = Xorb(o_out,isave) = contd5(0,xout);
      pv[1] = Yorb(o_out,isave) = contd5(1,xout);
      pv[2] = Zorb(o_out,isave) = contd5(2,xout);
      pv[3] = Uorb(o_out,isave) = contd5(3,xout);
      pv[4] = Vorb(o_out,isave) = contd5(4,xout);
      pv[5] = Worb(o_out,isave) = contd5(5,xout);
      (void) print_diag(xout, pv);
      xout += dtout;
    }
  }
}


void integrate_dopri5(void)
{
  int i, ndim, kdiag, ksave, isave, res, iout, itoler;
  double time, rtoler, atoler;
  double y[6];
  
  dprintf (1,"DOPRI5 integration\n");
  /* take last step of input file and set first step for outfile */
  time = Torb(o_out,0) = Torb(o_in,Nsteps(o_in)-1);
  y[0] = Xorb(o_out,0) = Xorb(o_in,Nsteps(o_in)-1);
  y[1] = Xorb(o_out,0) = Yorb(o_in,Nsteps(o_in)-1);
  y[2] = Zorb(o_out,0) = Zorb(o_in,Nsteps(o_in)-1);
  y[3] = Uorb(o_out,0) = Uorb(o_in,Nsteps(o_in)-1);
  y[4] = Vorb(o_out,0) = Vorb(o_in,Nsteps(o_in)-1);
  y[5] = Worb(o_out,0) = Worb(o_in,Nsteps(o_in)-1);
  
  ndim=Ndim(o_in);		/* number of dimensions (2 or 3) */
  
  iout = 2;     /* controlling solout */
  itoler = 0;
  rtoler = eta;
  atoler = rtoler;
  res = dopri5(6, rhs, 0.0, y, tstop, &rtoler, &atoler, itoler, solout5, iout,
	       stdout, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 6, NULL, 
	       3);
}

void solout8(long nr, double xold, double x, double *y, unsigned n, int *irtrn)
{
  static double xout;
  static int isave;
  double pv[6];

  if (nr==1) {
    xout = x + dtout;
    isave = 0;
    Torb(o_out,isave) = x;
    Xorb(o_out,isave) = y[0]; Yorb(o_out,isave) = y[1];  Zorb(o_out,isave) = y[2];
    Uorb(o_out,isave) = y[3]; Vorb(o_out,isave) = y[4];  Worb(o_out,isave) = y[5];
#ifdef ORBIT_PHI
    // steal y[6], 7,8,9, for phi/acc ?
    // Porb(o_out,isave) =
    // AXorb(o_out,isave) = 
    // AYorb(o_out,isave) = 
    // AZorb(o_out,isave) = 
#endif      
    (void) print_diag(x, y);
  } else  {
    while (x >= xout) {
      isave += 1;
      Torb(o_out,isave) = xout;
      pv[0] = Xorb(o_out,isave) = contd8(0,xout);
      pv[1] = Yorb(o_out,isave) = contd8(1,xout);
      pv[2] = Zorb(o_out,isave) = contd8(2,xout);
      pv[3] = Uorb(o_out,isave) = contd8(3,xout);
      pv[4] = Vorb(o_out,isave) = contd8(4,xout);
      pv[5] = Worb(o_out,isave) = contd8(5,xout);
      (void) print_diag(xout, pv);
      xout += dtout;
    }
  }
}

void integrate_dop853()
{
  int i, ndim, kdiag, ksave, isave, res, iout, itoler;
  double time, rtoler, atoler;
  double y[6];
  
  dprintf (1,"DOP853 integration\n");
  /* take last step of input file and set first step for outfile */
  time = Torb(o_out,0) = Torb(o_in,Nsteps(o_in)-1);
  y[0] = Xorb(o_out,0) = Xorb(o_in,Nsteps(o_in)-1);
  y[1] = Xorb(o_out,0) = Yorb(o_in,Nsteps(o_in)-1);
  y[2] = Zorb(o_out,0) = Zorb(o_in,Nsteps(o_in)-1);
  y[3] = Uorb(o_out,0) = Uorb(o_in,Nsteps(o_in)-1);
  y[4] = Vorb(o_out,0) = Vorb(o_in,Nsteps(o_in)-1);
  y[5] = Worb(o_out,0) = Worb(o_in,Nsteps(o_in)-1);
  
  ndim=Ndim(o_in);		/* number of dimensions (2 or 3) */
  
  iout = 2;     /* controlling solout */
  itoler = 0;
  rtoler = eta;
  atoler = rtoler;
  res = dop853(6, rhs, 0.0, y, tstop, &rtoler, &atoler, itoler, solout8, iout,
	       stdout, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 6, NULL, 
	       3);
}


/*
 *	PRINT_DIAG: print diagnostics, also add centrifugal term
 *		    returns the total energy in the rotating frame
 */

real print_diag(double time, double *posvel)
{
    double ekin;
    double epot;
    permanent bool first = TRUE;
    permanent double etot_0, etot_00;
    permanent int idiag = 0;
    double err;
    double f[6];
    double *pos = posvel;
    double *vel = pos+3;

    rhs(6,time,posvel,f);
    epot = last_pot;
	
    ekin=sqr(vel[0]) + sqr(vel[1])+ sqr(vel[2]);
    ekin *= 0.5;
    epot -= 0.5*omega2*(sqr(pos[0]) + sqr(pos[1])+ sqr(pos[2]));
    if (first) {
        dprintf (1,"time   Etot =   ekin +  epot\n");
	first = FALSE;
	etot_0 = epot + ekin;
        etot_00 = (etot_0 == 0.0 ? 1.0 : etot_0);
    }
    err = (epot+ekin-etot_0)/etot_00;
    if (ndiag) {
      if (idiag%ndiag == 0)
	dprintf(0,"%f %f %f %20.13g %g\n",time,ekin,epot,ekin+epot,err);
    }
    idiag++;
    return ekin+epot;
}
